packages/venia-concept/src/components/App/errorNotifications.js

```diff
@@ -0,0 +1,57 @@
+import React, { Component } from 'react';
+import { arrayOf, object, func, shape, string } from 'prop-types';
+import classify from 'src/classify';
+import { Notification, NotificationStack } from 'src/components/Notifications';
+import defaultClasses from './errorNotifications.css';
+
+class ErrorNotifications extends Component {
+    static propTypes = {
+        classes: shape({
+            debuginfo: string
+        }).isRequired,
+        errors: arrayOf(
+            shape({
+                error: object.isRequired,
+                id: string.isRequired,
+                loc: string
+            })
+        ),
+        onDismissError: func.isRequired
+    };
+
+    componentDidMount() {
+        window.scrollTo(0, 0);
+    }
+
+    render() {
+        const { classes, onDismissError, errors } = this.props;
+        if (errors.length > 0) {
+            return (
+                <div className={classes.root}>
+                    <NotificationStack>
+                        {errors.map(({ error, id, loc }) => (
+                            <Notification
+                                key={id}
+                                type="error"
+                                onClick={(e, dismiss) => {
+                                    e.preventDefault();
+                                    e.stopPropagation();
+                                    dismiss();
+                                }}
+                                afterDismiss={() => onDismissError(error)}
```

Comment:
These handlers need to be defined as class members or memoized with `useCallback`. As it is, they're being recreated and rebound on every render.

Reply:
Since renders of these should be rare, I'd like to leave this. It's harder than it looks to change.

- Can't use hooks in a class component: https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both
- The inline function allows access to the `error` object in closure. The Notification doesn't know about the error object, so it can't pass it back as an argument to afterDismiss.

Reply:
> The inline function allows access to the `error` object in closure.     * The Notification doesn't know about the error object...

I understand—this is a common circumstance, but it's still a problem that needs to be addressed. There are solutions that don't involve major refactoring, though. One of the following things needs to happen:

- `Notification` needs to receive `error` as a prop and define the handler
- An intermediate component needs to receive `error` and define the handler
- The handler needs to be memoized

```es6
const handlers = new WeakMap()

const handler = handlers.has(error)
  ? handlers.get(error)
  : handlers.set(error, () => { log(error); }).get(error)
```